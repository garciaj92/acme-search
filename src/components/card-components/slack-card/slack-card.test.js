import React from "react";
import SlackCard from "./slack-card";

import { mount } from "enzyme";
import toJson from 'enzyme-to-json';

// Mocks
import { consoleErrorFn } from "../../../__mocks__/console.mock";

const props = {
    channel: "test-channel",
    author: "jorgelgarcia",
    message: "This is a test.",
    createdAt: "01/01/2020 - 9:00:00 AM"
};

describe("SlackCard: Test visual components", () => {
    let wrapper;
    beforeEach(() => {
        wrapper = mount(<SlackCard {...props} />);
    });

    it("renders correctly", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
        // console.log(wrapper.debug());
    });

    it("displays the props correctly", () => {
        expect(wrapper.find("[data-testid='card-heading']").first().text()).toEqual(props.author);
        expect(wrapper.find("[data-testid='card-info'] h5").at(0).text()).toEqual("Channel");
        expect(wrapper.find("[data-testid='card-info'] p").at(0).text()).toEqual(props.channel);
        expect(wrapper.find("[data-testid='card-info'] h5").at(1).text()).toEqual("Message");
        expect(wrapper.find("[data-testid='card-info'] p").at(1).text()).toEqual(props.message);
        expect(wrapper.find("[data-testid='card-footer'] h5").first().text()).toEqual("Sent");
        expect(wrapper.find("[data-testid='card-footer'] p").first().text()).toEqual(props.createdAt);
    });
});


describe("SlackCard: Test prop type validation", () => {
    let consoleErrorSpy;
    beforeEach(() => {
        console.error = consoleErrorFn
        consoleErrorSpy = jest.spyOn(console, "error");
    });

    afterEach(() => {
        consoleErrorSpy.mockClear();
    });

    it("does not throw an error when props are provided", () => {
        mount(<SlackCard {...props} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    // What is stated below applies to the remaining test cases in this describe block:
    //      console.error is called twice for each prop validator that fails because
    //      there are two components that use each failing prop. This means that if
    //      two props fail we have four errors.

    it("throws an error for each prop that is not provided", () => {
        mount(<SlackCard />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(8);
    });

    it("throws an error when author prop is not of type `string`", () => {
        const { channel, message, createdAt } = props;
        mount(<SlackCard author={1} channel={channel} message={message} createdAt={createdAt} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when channel prop is not of type `string`", () => {
        const { author, message, createdAt } = props;
        mount(<SlackCard author={author} channel={1} message={message} createdAt={createdAt} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(3);
    });

    it("throws an error when message prop is not of type `string`", () => {
        const { channel, author, createdAt } = props;
        mount(<SlackCard author={author} channel={channel} message={1} createdAt={createdAt} />);

        // SPECIAL CASE
        // The spy would normally be called 2 times but since slack-card uses
        // card-info twice and card-info has already previously thrown an error
        // there is only one error message thrown for the missing content prop
        expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
    });

    it("throws an error when createdAt prop is not of type `string`", () => {
        const { channel, author, message } = props;
        mount(<SlackCard author={author} channel={channel} message={message} createdAt={1} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });
});