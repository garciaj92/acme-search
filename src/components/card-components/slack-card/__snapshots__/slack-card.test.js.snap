// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlackCard: Test visual components renders correctly 1`] = `
<SlackCard
  author="jorgelgarcia"
  channel="test-channel"
  createdAt="01/01/2020 - 9:00:00 AM"
  message="This is a test."
>
  <Card
    bg="white"
    body={false}
    className="shadow calendar-card"
  >
    <div
      className="shadow calendar-card card bg-white"
    >
      <CardBody>
        <div
          className="card-body"
        >
          <CardHeading
            data-testid="card-heading"
            heading="jorgelgarcia"
          >
            <h4
              className="font-weight-bold"
            >
              jorgelgarcia
            </h4>
          </CardHeading>
          <CardInfo
            content="test-channel"
            data-testid="card-info"
            heading="Channel"
            iconSrc="slack-hash.svg"
          >
            <CardIconSubHeading
              heading="Channel"
              iconSrc="slack-hash.svg"
            >
              <div
                className="d-flex justify-content-center align-items-center my-2"
              >
                <Image
                  fluid={false}
                  rounded={false}
                  roundedCircle={false}
                  src="slack-hash.svg"
                  thumbnail={false}
                  width={30}
                >
                  <img
                    className=""
                    src="slack-hash.svg"
                    width={30}
                  />
                </Image>
                <h5
                  className="mb-0 mx-1"
                >
                  Channel
                </h5>
              </div>
            </CardIconSubHeading>
            <CardLink
              content="test-channel"
            >
              <p
                className="text-muted mb-0 d-block"
              >
                test-channel
              </p>
            </CardLink>
          </CardInfo>
          <CardInfo
            content="This is a test."
            data-testid="card-info"
            heading="Message"
            iconSrc="message-circle.svg"
          >
            <CardIconSubHeading
              heading="Message"
              iconSrc="message-circle.svg"
            >
              <div
                className="d-flex justify-content-center align-items-center my-2"
              >
                <Image
                  fluid={false}
                  rounded={false}
                  roundedCircle={false}
                  src="message-circle.svg"
                  thumbnail={false}
                  width={30}
                >
                  <img
                    className=""
                    src="message-circle.svg"
                    width={30}
                  />
                </Image>
                <h5
                  className="mb-0 mx-1"
                >
                  Message
                </h5>
              </div>
            </CardIconSubHeading>
            <CardLink
              content="This is a test."
            >
              <p
                className="text-muted mb-0 d-block"
              >
                This is a test.
              </p>
            </CardLink>
          </CardInfo>
        </div>
      </CardBody>
      <CardFooter
        className="bg-white"
        data-testid="card-footer"
      >
        <div
          className="bg-white card-footer"
          data-testid="card-footer"
        >
          <DateInfo
            date="01/01/2020 - 9:00:00 AM"
            title="Sent"
          >
            <div
              className="text-muted"
            >
              <h5
                className="font-weight-bold mb-0"
              >
                Sent
              </h5>
              <p
                className="font-weight-bold mb-0"
              >
                01/01/2020 - 9:00:00 AM
              </p>
            </div>
          </DateInfo>
        </div>
      </CardFooter>
    </div>
  </Card>
</SlackCard>
`;
